### DISCUSSION TITLE
Algorithm for Set Operations

### DISCUSSION THREAD

**Discussion by EricZimmermann** [shapely/shapely#1813] [Q&A] [answered]
Hello,

I'm looking to find the algorithm for implementing the sym difference between two polygons but I can't seem to find the src file which implements it. I'm trying to understand how these pieces are computed. 

Thank you!

**Comment by jorisvandenbossche** [comment, marked as answer]
The actual algorithm is actually implemented in the GEOS library, as shapely's function is essentially a small python binding to `GEOSSymDifference_r`. 

I don't know if there is a single place to point to where this is implemented, but the C API function we call is essentially defined here: https://github.com/libgeos/geos/blob/4c3bd72ba73e9688b0712d9878d073559c84e4d6/src/geom/Geometry.cpp#L570

**Comment by mwtoews** [comment]
The actual algorithm is done by GEOS' [OverlayNG](https://libgeos.org/doxygen/classgeos_1_1operation_1_1overlayng_1_1OverlayNG.html) done in C++ which was ported from JTS' [OverlayNG](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/operation/overlayng/OverlayNG.html) done in Java. See also blog posts [here](http://lin-ear-th-inking.blogspot.com/2020/05/jts-overlay-next-generation.html) or [here](http://lin-ear-th-inking.blogspot.com/2020/06/jts-overlayng-noding-strategies.html), or [other posts labeled with "overlay"](http://lin-ear-th-inking.blogspot.com/search/label/overlay). The actual algorithm is not easy to "see", so most users (including myself) just have faith that they somehow work!

**Comment by EricZimmermann** [comment]
Thanks everyone! This should solve my issues! I am assuming I can approximate these differences by rasterizing my polygons and computing the difference with binary maps. The idea of snapping means I can estimate my error to arbitrary precision (at the cost of memory) by selecting the resolution of a grid to rasterize onto :)

**Reply by mwtoews** [reply]
There is also a [`grid_size` parameter](https://shapely.readthedocs.io/en/stable/reference/shapely.symmetric_difference.html) to snap geometries to a grid, which has the same effect as rasterizing methods. But raster methods are fine too!
